Digital Credential: provide some means to validate a request

marcoscaceres opened the pull request:
#### 55b96329e33bc06f2573303c721f07312aa1d01a
<pre>
Digital Credential: provide some means to validate a request
<a href="https://bugs.webkit.org/show_bug.cgi?id=286660">https://bugs.webkit.org/show_bug.cgi?id=286660</a>
<a href="https://rdar.apple.com/143799838">rdar://143799838</a>

Reviewed by Richard Robinson.

This patch adds support for ISO18013-5 Mobile Document format in WebKit, including infrastructure to allow validations of such requests.

Adds a notion of validated and unvalidated requests, which allows the system to handle both validated and unvalidated requests in a more structured way.

The CredentialRequestCoordinator class:

- Becomes ActiveDOMObject, allowing it to monitor for frames being destroyed. This allows for a more robust handling of the lifecycle of the credential request and associated promises.

The DigitalCredentialsCoordinator class:

- Becomes a message receiver, allowing clients to request raw data only when needed via provideRawDigitalCredentialRequests()

Validation can now be handled generically via a DigitalCredentialsRequestValidatorBridge, which provides a bridge between the WebKit and WebCore layers.

Canonical link: <a href="https://commits.webkit.org/294746@main">https://commits.webkit.org/294746@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/99b09eb89f9c3a15f8584c813ec07fec4aa34e57

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/102766 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/22439 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/12759 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/107931 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/53407 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/104805 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/22756 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/30941 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/78147 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/35113 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/105772 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/17624 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/92743 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/58479 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/17469 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/10778 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/52764 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/87280 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/10846 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/110307 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/29903 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/22036 "Passed tests") | [⏳ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/GTK-WK2-Tests-EWS "Waiting to run tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/30267 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/88939 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/86742 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22109 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/31579 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/9302 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/24159 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/29830 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/35151 "Failed to build and analyze WebKit") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/29638 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/32965 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/31200 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/Platform/cocoa/CocoaWindow.h:
```diff
@@ -27,8 +27,8 @@
 #else
 OBJC_CLASS UIWindow;
-using CocoaWindow = UIWindow;
+typedef UIWindow CocoaWindow;
```
I am surprised by this change, we've been replacing typedefs with `using` statements for a while now so this seems to be going in the opposite direction to the rest of the codebase.

cdumez commented on Source/WebCore/Modules/identity/CredentialRequestCoordinator.cpp:
```diff
@@ -159,22 +175,33 @@ void CredentialRequestCoordinator::presentPicker(CredentialPromise&& promise, Di
+ExceptionOr<DigitalCredentialsRequestData> CredentialRequestCoordinator::prepareDigitalCredentialRequestData(const Document& document, Vector<ValidatedDigitalCredentialRequest>&& requests)
+{
+    DigitalCredentialsRequestData requestData;
+    requestData.topOrigin = document.protectedTopOrigin()->data().isolatedCopy();
```
Why the isolatedCopy()?

marcoscaceres replied:
copy/pasta... I'll change it.

marcoscaceres replied:
Actually, I used that because it said:
```
    // Create a deep copy of this SecurityOrigin. This method is useful
    // when marshalling a SecurityOrigin to another thread.
    WEBCORE_EXPORT Ref<SecurityOrigin> isolatedCopy() const;
```

Is `crossThreadCopy(document.protectedTopOrigin()->data())` the right thing to do here?

cdumez replied:
It is not clear to me how multi-threading is involved here. Normally we call `crossThreadCopy()` / `isolatedCopy()` when dispatching to another thread. Calling them inside a function like this that is not named isolatedCopy() and doesn't dispatch to another thread is unexpected, thus my comment.

marcoscaceres replied:
Right, I'd misunderstood the intent of `.isolatedCopy()`. Hopefully all good now.

aprotyas commented on Source/WebCore/Modules/identity/DigitalCredential.h:
```diff
@@ -64,6 +69,9 @@ class DigitalCredential final : public BasicCredential {
 
+    static ExceptionOr<Vector<ValidatedDigitalCredentialRequest>> validateRequests(const Document&, Vector<UnvalidatedDigitalCredentialRequest>&&);
+    static ExceptionOr<Vector<UnvalidatedDigitalCredentialRequest>> convertObjectsToDigitalPresentationRequests(const Document&, const Vector<DigitalCredentialRequest>&);
+    static ExceptionOr<DigitalCredentialsRequestData> prepareDigitalCredentialRequestData(const Document&, Vector<ValidatedDigitalCredentialRequest>&&);
```
I think you can remove this declaration, since it seems to be undefined, and you use the equivalent provided by `CredentialRequestCoordinator`.

rr-codes commented on Source/WebCore/Modules/identity/CredentialRequestCoordinator.h:
```diff
@@ -27,16 +27,17 @@
 
 namespace WebCore {
-
+class BasicCredential;
```
this should be below AbortSignal for alphabetical ordering

rr-codes commented on Source/WebCore/Modules/identity/protocols/DigitalCredentialsProtocols.h:
```diff
@@ -0,0 +1,37 @@
+
+namespace WebCore {
+
+using ValidatedDigitalCredentialRequest = std::variant<ValidatedMobileDocumentRequest, OpenID4VPRequest>;
```
WTF::Variant

rr-codes commented on Source/WebKit/WebKitSwift/IdentityDocumentServices/WKISO18013Request.h:
```diff
@@ -0,0 +1,37 @@
+
+@interface WKISO18013Request : NSObject
+
+@property (nonatomic, strong) NSString *encryptionInfo;
```
no need for strong since its the default

pxlcoder replied:
Strings should normally be `copy`.

rr-codes commented on Source/WebKit/WebKitSwift/IdentityDocumentServices/WKIdentityDocumentPresentmentDelegate.h:
```diff
@@ -0,0 +1,48 @@
+#endif
+
+NS_ASSUME_NONNULL_BEGIN
+NS_SWIFT_UI_ACTOR
```
this should be NS_ASSUME_NONNULL_BEGIN , then an empty line, and then NS_SWIFT_UI_ACTOR, and then no empty line

rr-codes commented on Source/WebKit/WebKitSwift/WebKitSwift.h:
```diff
@@ -34,3 +34,11 @@
 #import "WKSPreviewWindowController.h"
 #import "WKSTextAnimationSourceDelegate.h"
 #import "WKStageMode.h"
+#import "WKIdentityDocumentPresentmentController.h"
```
alphabetical order

rr-codes commented on Source/WebKit/WebProcess/cocoa/IdentityDocumentServices/DigitalCredentialsRequestValidatorBridge.mm:
```diff
@@ -0,0 +1,179 @@
+
+namespace WebKit {
+
+using namespace WebCore;
```
should generally avoid `using namespace`

pxlcoder commented on Source/WebCore/Modules/identity/protocols/ISO18013/ISO18013ElementInfo.h:
```diff
@@ -0,0 +1,34 @@
+namespace WebCore {
+
+struct ISO18013ElementInfo {
+    bool isRetaining = true;
```
Nit: WebKit style is normally

```suggestion
    bool isRetaining { true };
```

pxlcoder commented on Source/WebKit/UIProcess/DigitalCredentials/WKDigitalCredentialsPicker.mm:
```diff
@@ -174,7 +182,6 @@ - (void)completeWith:(Expected<WebCore::DigitalCredentialsResponseData, WebCore:
     _completionHandler = nullptr;
     [self dismiss];
 }
-
```
Undo.

pxlcoder commented on Source/WebKit/WebProcess/cocoa/IdentityDocumentServices/DigitalCredentialsRequestValidatorBridge.mm:
```diff
@@ -0,0 +1,179 @@
+
+using namespace WebCore;
+
+static RetainPtr<SecTrustRef> createSecTrustForChain(const Vector<RetainPtr<SecCertificateRef>> &chain)
```
```suggestion
static RetainPtr<SecTrustRef> createSecTrustForChain(const Vector<RetainPtr<SecCertificateRef>>& chain)
```

marcoscaceres replied:
clang-format keeps changing this :( ... I'll change it, but maintaining whitespace manually is not fun 😿

pxlcoder commented on Source/WebKit/WebProcess/cocoa/IdentityDocumentServices/DigitalCredentialsRequestValidatorBridge.mm:
```diff
@@ -0,0 +1,179 @@
+
+    RetainPtr cfChain = adoptCF(CFArrayCreateMutable(kCFAllocatorDefault, 0, &kCFTypeArrayCallBacks));
+
+    for (auto &cert : chain)
```
We should be explicit about the smart point here rather than use `auto`.

marcoscaceres committed:
Digital Credential: provide some means to validate a request
https://bugs.webkit.org/show_bug.cgi?id=286660
rdar://143799838

Reviewed by Richard Robinson.

This patch adds support for ISO18013-5 Mobile Document format in WebKit, including infrastructure to allow validations of such requests.

Adds a notion of validated and unvalidated requests, which allows the system to handle both validated and unvalidated requests in a more structured way.

The CredentialRequestCoordinator class:

- Becomes ActiveDOMObject, allowing it to monitor for frames being destroyed. This allows for a more robust handling of the lifecycle of the credential request and associated promises.

The DigitalCredentialsCoordinator class:

- Becomes a message receiver, allowing clients to request raw data only when needed via provideRawDigitalCredentialRequests()

Validation can now be handled generically via a DigitalCredentialsRequestValidatorBridge, which provides a bridge between the WebKit and WebCore layers.

Canonical link: https://commits.webkit.org/294746@main
